src/Service.Tests/dab-config.MsSql.json

```diff
@@ -3674,6 +3674,58 @@
           "linking.target.fields": []
         }
       }
+    },
+    "dbo_DimAccount": {
+      "source": {
+        "object": "DimAccount",
+        "type": "table"
+      },
+      "graphql": {
+        "enabled": true,
+        "type": {
+          "singular": "dbo_DimAccount",
+          "plural": "dbo_DimAccounts"
+        }
+      },
+      "rest": {
+        "enabled": true
+      },
+      "permissions": [
+        {
+          "role": "anonymous",
+          "actions": [
+            {
+              "action": "*"
+            }
+          ]
+        }
+      ],
+      "relationships": {
+        "parent_account": {
+          "cardinality": "one",
+          "target.entity": "dbo_DimAccount",
```

Comment:
From the cardinality, how can you infer that the source is `ReferencingEntity` for sure? I get this is possible in many:one relationship. 

How about one:one relationships? In those either side can be the referencing entity, no? depending upon which entity contains the foreign key..

Reply:
For queries, we do not need to accurately identify the referencing and referencing entities. 

For relationships with cardinality `one`, DAB adds foreign key definition in both entities. Eventually, this translates to an additional incorrect condition during the join operation. This extra condition is the cause for returning no results getting returned (as both the conditions are never satisfied). 

When the foreign key definition is added in both entities, the database query is of the form,
```sql
WHERE [table0].[ParentAccountKey] = [table1].[AccountKey]
      AND [table0].[AccountKey]       = [table1].[ParentAccountKey]
```

but, we would want only the right equality check.

```sql
WHERE [table0].[ParentAccountKey] = [table1].[AccountKey]
```
Generating the right condition is only dependent on identifying and matching the top-level entity and the nested entity with their corresponding relationship fields. 

So, essentially, for self-join, 
-  Foreign key definition should be populated only with one combination (source --> referencing, target --> referenced or vice-versa. This doesn't have to be accurate) to avoid the extra incorrect condition. 
- `ForeignKeyDefinition` data structure contains `ReferencingColumns`, `ReferencedColumns` and `RelationShipPair`. We do not immediately know which is the referenced table and match it with referenced columns (same for referencing columns) when generating the sql query. In different places, the database object is compared to see if it is the referencedtable/referencingtable before fetching referencing/referenced columns. With self-join relationships, this is a challenge, because the database object is the same for referencing and referenced tables, hence the need for relationship roles and helper methods such as `ResolveSourceColumns()`, `ResolveTargetColumns()` which enable matching the top-level entity/nested entity with their associated relationship fields.

Reply:
I understand this reasoning. What confused me is the fact that the description doesn't say, we need to use source entity as referencing + target entity as referenced when cardinality is `one` ONLY when the db doesn't have a foreign key defined in it.